### DISCUSSION TITLE
Is it possible to create SSG Components?

### DISCUSSION THREAD

**Discussion by EliasSchlatter** [vercel/next.js#18862] [Help] [answered]
Hi everybody,

### Description
We have a SSG page whose main content is displayed in the middle of the screen. On the left side there is a Drawer menu with links to other exercises (it is an e-learning site). To display these links data has to be loaded from the backend. 

If we now use a "revalidate" to update exercises, a lot of pages have to be re-rendered because a small part of their drawer has changed.

### What we would like achieve
It would be perfect if the Drawer and the exercise could be pre-rendered independently.

### Case Example
Imagine there are 40 links to different exercises in the Drawer. If now another exercise is added, all 40 existing exercises + the newly created exercise have to be re-rendered, because a part of the Drawer has changed.

What we would like to achieve is that only the new exercise has to be rendered + 1 time the Drawer, which then can be used again in all exercises that have been there before.

Thanks for your help. :)

**Comment by lfades** [comment, marked as answer]
@EliasSchlatter What you're asking for is very hard to get done, prerendered pages are static HTML created at build time, if you have 1000 pages using the same Drawer, those pages are not including a "shared Drawer component" but the HTML output of that component, so if you ever change the Drawer component all of your pages will be out of date because updating a component won't update the static HTML in all other pages. Having said that, don't be afraid of prerendering a lot of pages again, that's actually okay! Next.js can pre-render thousands of pages very fast and we're actively working on making that even faster. And if you actually have an incredible amount of pages, there's always the option of using [fallback pages](https://nextjs.org/docs/basic-features/data-fetching#fallback-pages) and render the content in the background when a request comes in.

**Comment by EliasSchlatter** [comment]
Hi @lfades, thanks for the quick response!

I would love to simply re-rendered all affected pages of a drawer change, but unfortunately it takes a very long time, sometimes more than 20 minutes, until a change is visible.

However, I am currently running the app on rather weak computers. To be precise a Kubernetes cluster with 3 * n1-standard-2 (Google Cloud) nodes, which have a total of six cores and 22.50 GB memory. Backend and a MongoDB run on the same cluster. Could this be the source of the problem? If so, can you give me a recommendation on how powerful the running machines should be?

Also I have seen that when I start a build requests that are the same on every page are sent over and over again. Is this also the case during the revalidate? If so, could too large request load be the reason for slow loading. At the moment about 60 KB must be loaded to render a page.

Can you think of another reason why rendering takes so long?

PS: regarding the fallback option, if I turn it to true on the page with the most number of rerenderings, it still takes quiet some time to rerender all other affected pages. Also I have seen, that a Page, that has fallback set to false is not updated for a long time. Does this perhaps indicate that something about the validation of whether something needs to be re-rendered does not work at all?

**Reply by lfades** [reply]
Knowing why build times take too long is hard to figure out without looking at the app, so here are my recommendations:

- Look at how are the build times in http://vercel.com/, if they're better then you may want to improve your setup (also compare it with your local builds if you have a good machine).
- [Enable webpack 5](https://nextjs.org/blog/next-9-5#webpack-5-support-beta), is experimental but it may improve your build times.
- See if there is any expensive tasks that may be increasing a lot the build times, for comparison all pages in http://vercel.com build between 7 to 9 minutes

**Reply by JonCognioDigital** [reply]
Isn't there another issue here, which is that some of those pages will re-render almost immediately because they haven't been generated for a while but one pages will not because they are still within the revalidate time.

This means some pages will have the new menu and some won't.

We're not using SSG in our latest project as it just seems too flakey at the moment. Maybe I just don't understand it well enough yet but there seem to be lots of issues with caching and stale content

**Reply by lfades** [reply]
@jon64digital We pregenerate all pages at build time, so if a component changes it will be reflected in all pages, however if the contents of a page changes at runtime with revalidation, not all pages will be revalidated so you may see different content for pages that have not been accessed in some time (until there's a new build that is). We'll eventually work in something like layouts to improve the way common data is shared across pages.

**Reply by JonCognioDigital** [reply]
"We'll eventually work in something like layouts to improve the way common data is shared across pages."

This sounds like it would be a massive improvement and solve one of the major problems with SSG. Are there any official plans for this yet?

**Reply by lfades** [reply]
@jon64digital We're currently focused on bug fixes and improvements for the new features released in Next.js 10, once that's done we'll start working in new features and layouts may be one of those, but nothing official for now.

**Reply by devtim123** [reply]
@lfades are there any updates on this? would be very nice and sustainable... something like block and tag based caching for components would be awesome...
